pkg/storegateway/indexheader/index/postings.go

```diff
@@ -87,58 +87,105 @@ func newV1PostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset
 	return &t, nil
 }
 
-func newV2PostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64, postingOffsetsInMemSampling int) (*PostingOffsetTableV2, error) {
+type lastValuePosition struct {
+	key                string
+	valueOffsetFromEnd int
+	offsetInTable      int
+}
+
+func newV2PostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64, postingOffsetsInMemSampling int) (table *PostingOffsetTableV2, err error) {
 	t := PostingOffsetTableV2{
 		factory:                     factory,
 		tableOffset:                 tableOffset,
 		postings:                    map[string]*postingValueOffsets{},
 		postingOffsetsInMemSampling: postingOffsetsInMemSampling,
 	}
 
-	lastTableOff := 0
-	valueCount := 0
-	var lastKey string
-	var lastValue string
+	d := factory.NewDecbufAtChecked(tableOffset, castagnoliTable)
+	defer runutil.CloseWithErrCapture(&err, &d, "read offset table")
 
-	// For the postings offset table we keep every label name but only every nth
-	// label value (plus the first and last one), to save memory.
-	if err := readOffsetTable(factory, tableOffset, func(key string, value string, off uint64, tableOff int) error {
-		if _, ok := t.postings[key]; !ok {
-			// Not seen before label name.
-			if len(t.postings) > 0 {
-				// Always include last value for each label name, unless it was just added in previous iteration based
-				// on valueCount.
-				if (valueCount-1)%postingOffsetsInMemSampling != 0 {
-					t.postings[lastKey].offsets = append(t.postings[lastKey].offsets, postingOffset{value: lastValue, tableOff: lastTableOff})
-				}
+	startLen := d.Len()
+	remainingCount := d.Be32()
+	currentKey := ""
+	valuesForCurrentKey := 0
+
+	var lastValuePositions []lastValuePosition
+
+	for d.Err() == nil && remainingCount > 0 {
+		lastKey := currentKey
+		offsetInTable := startLen - d.Len()
+		keyCount := d.Uvarint()
+
+		// The Postings offset table takes only 2 keys per entry (name and value of label).
+		if keyCount != 2 {
+			return nil, errors.Errorf("unexpected key length for posting table %d", keyCount)
+		}
+
+		// Important: this value is only valid as long as we don't perform any further reads from d.
+		// If we need to retain its value, we must copy it before performing another read.
+		key := d.UvarintBytes()
+
+		if len(t.postings) == 0 || currentKey != string(key) {
+			// New label name.
+			currentKey = string(key)
+			t.postings[currentKey] = &postingValueOffsets{}
+			lastValuePositions = append(lastValuePositions, lastValuePosition{key: currentKey, valueOffsetFromEnd: -1})
+			valuesForCurrentKey = 0
+		}
+
+		if valuesForCurrentKey%postingOffsetsInMemSampling == 0 {
+			value := d.UvarintStr()
+			off := d.Uvarint64()
+			t.postings[currentKey].offsets = append(t.postings[currentKey].offsets, postingOffset{value: value, tableOff: offsetInTable})
+
+			if lastKey != currentKey {
```

Comment:
Can you explain me this, please? Adding a comment may help. I got lost here (I think I got lost even when I reviewed the initial mmap-less PR).

Reply:
Are you asking about `int64(off - crc32.Size)`? Or why we need `lastValOffset`?

Reply:
I'm asking about the whole `if` block:

```
if lastName != currentName {
    t.postings[lastName].lastValOffset = int64(off - crc32.Size)
}
```

If we switched to a new symbol name, we update the previous symbol name setting the offset of the last value to `off - crc32.Size`. But isn't `off` the offset of the current symbol value? I'm sure I'm missing something...

Reply:
I'm not 100% sure about this, but this is my understanding:

`t.postings[name].lastValOffset` is the offset of the last byte of the [postings section](https://github.com/prometheus/prometheus/blob/main/tsdb/docs/format/index.md#postings) in the index file for the last value for name `name`. We use this to calculate what range of the index file we need to download if a query touches this label.

Postings sections are sorted in the same order used to sort the postings offset table. So if we've just read the first postings offset table entry for a new label name, and this first entry of the new label has its postings section at offset `off`, then the previous postings section (for the last value for the previous name) ends just before `off`. 

Each postings section ends with a CRC32 checksum, and we don't check these at query time, so we can save 4 bytes by skipping the trailing four bytes.

Visually, it looks like this:


```
offset for "name-1" >┌────────────────────┬────────────────────┐
="value-9" points    │ len <4b>           │ #entries <4b>      │
here                 ├─┬──────────────────┼──────────────────┬─┤
                     │ ├─────────────────────────────────────┤ │
                     │ │ ref(series_1) <4b>                  │ │
                     │ ├─────────────────────────────────────┤ │  Postings
                     │ │ ...                                 │ │  section for
                     │ ├─────────────────────────────────────┤ │  "name-1"="value-9"
                     │ │ ref(series_n) <4b>                  │ │
                     │ ├─────────────────────────────────────┤ │
                     ├─┴─────────────────────────────────────┴─┤
                     │ CRC32 <4b>                              │
offset for "name-2" >├────────────────────┬────────────────────┤
="value-1" points    │ len <4b>           │ #entries <4b>      │
here                 ├─┬──────────────────┼──────────────────┬─┤
                     │ ├─────────────────────────────────────┤ │
                     │ │ ref(series_1) <4b>                  │ │
                     │ ├─────────────────────────────────────┤ │  Postings
                     │ │ ...                                 │ │  section for
                     │ ├─────────────────────────────────────┤ │  "name-2"="value-1"
                     │ │ ref(series_n) <4b>                  │ │
                     │ ├─────────────────────────────────────┤ │
                     ├─┴─────────────────────────────────────┴─┤
                     │ CRC32 <4b>                              │
                     └─────────────────────────────────────────┘

                                          .
                                          .
                                          .
                                          .
```